### PR TITLE
feat(graph-gateway): add network service

### PR DESCRIPTION
**Pull request by LNSD**
These PR changes aim to introduce a new network subgraph resolution service. The idea is to replace the current _types-eventuals-iterators_ code 🍝 with an encapsulated implementation.

- [x] Make the graph network subgraph topology info, pre-processing, and processing (e.g., indexer blocklist-based filtering) deterministic. The update fetch is awaited for a period, for example, 30 seconds. If the update fetch fails or takes longer than the update interval, a warning is logged, and a new fetch starts.
- [x] Resolve all indexers information in parallel (see the `futures::future::join_all()` part) with timeouts. This prevents an indexer from dragging the gateway with a slow response.
- [x] Rework the data structures to fit the `client_query.rs` requirements. For example, the data structure no longer follows the fetched subgraphs data structure. Now we have `Indexing` (and `IndexingId`) types associated with `SubgraphId` and `DeploymentId.`
- [x] Add integration tests that partially cover the functionality.

### CODE REVIEW

**Comment by LNSD** [comment]
NB: The modules are under the `graph-gateway` as the types and the API are coupled with the `client_query.rs` file logic.

File: graph-gateway/src/network/indexers_cost_model_resolver.rs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        indexer_deployments: &[DeploymentId],
+    ) -> HashMap<DeploymentId, CostModelSource> {
+        let sources =
+            match indexers::cost_models::query(&self.client, indexer_cost_url, indexer_deployments)
+                .await
+                .map_err(|err| ResolutionError::FetchFailed(err.to_string()))
+            {
+                Ok(sources) => sources,
+                Err(err) => {
+                    tracing::debug!("Failed to fetch cost models: {err}");
```
Comment (Theodus) [inline graph-gateway/src/network/indexers_cost_model_resolver.rs:47]: feels a bit redundant to have a message like `Failed to fetch cost models: The cost model couldn't be fetched: ...`

File: graph-gateway/src/network/entities.rs [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Ord, PartialOrd, Hash)]
+pub struct IndexingId {
+    /// The indexer's ID indexing the subgraph's deployment.
+    pub indexer: Address,
+    /// The subgraph's deployment ID indexed by the indexer.
+    pub deployment: DeploymentId,
+}
+
+#[derive(Clone)]
+pub struct Indexing {
```
Comment (Theodus) [inline graph-gateway/src/network/entities.rs:28]: I'm thinking that the term Indexing should be replaced with `Allocation`, and then we use the allocation id (`Address`) as keys in maps etc. And the `Allocation` struct would contain the corresponding indexer address, deployment id, etc.

Not necessary for this PR. Just a thought I had recently to remove this `Indexing` term that we've held over from the original TS gateway.

File: graph-gateway/src/network/entities.rs [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    /// The indexer's URL.
+    ///
+    /// It is guaranteed that the URL scheme is either HTTP or HTTPS and the URL has a host.
+    #[debug(with = Display::fmt)]
+    pub url: Url,
+
+    /// The total amount of tokens staked by the indexer.
+    pub staked_tokens: u128,
+    /// The largest allocation address.
+    pub largest_allocation: Address,
```
Comment (Theodus) [inline graph-gateway/src/network/entities.rs:72]: This field is concerning. The allocation should be associated with an indexing, not with the indexer.

File: graph-gateway/src/network/entities.rs [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    /// The total amount of tokens allocated for this indexer.
+    ///
+    /// It is the sum of all the `allocated_tokens` from the indexer's allocations.
+    pub total_allocated_tokens: u128,
```
Comment (Theodus) [inline graph-gateway/src/network/entities.rs:76]: What's the purpose of this field?

**Review by Theodus — COMMENTED** [review]
Still reviewing the last commit. Here's my comments so far.

File: graph-gateway/tests/it_network_subgraph_client.rs [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+fn hosted_service_url_with_subgraph_name(name: impl AsRef<str>) -> Url {
+    test_base_url()
+        .join(&format!("subgraphs/name/{}", name.as_ref()))
+        .expect("Invalid URL")
+}
+
+/// The Graph Network Mainnet in the hosted service.
+///
+/// https://thegraph.com/hosted-service/subgraph/graphprotocol/graph-network-mainnet
+const GRAPH_NETWORK_MAINNET_SUBGRAPH_ID: &str = "graphprotocol/graph-network-mainnet";
```
[the remark this answers is not in the document]
Comment (LNSD) [reply graph-gateway/tests/it_network_subgraph_client.rs:26]: I agree, but unfortunately, that would be unproductive at this moment. IIRC, our queries failed the last time I tested against the arbitrum graph-network subgraph because the subgraph schema was not up-to-date.

In any case, I aim to have a replacement for the current implementation in this PR. The tests can be changed to run against the arbitral subgraph by changing the env variable name.

Another possibility is to develop a solution based on cargo features that allow switching the target without duplicating the files using some `#[cfg(...)]` attributes.

File: graph-gateway/src/network/indexers_cost_model_resolver.rs (hunk 1 above)
Comment (LNSD) [reply graph-gateway/src/network/indexers_cost_model_resolver.rs:47]: I will review the error messages 👍🏼

File: graph-gateway/src/network/entities.rs (hunk 3 above)
Comment (LNSD) [reply graph-gateway/src/network/entities.rs:72]: I will review this 👍🏼

File: graph-gateway/src/network/entities.rs (hunk 4 above)
Comment (LNSD) [reply graph-gateway/src/network/entities.rs:76]: It is used to say if the indexing has a "zero-allocation". From `client_query.rs`:

```rust
Ok(Candidate {
    indexer: indexing.id.indexer,
    deployment: indexing.id.deployment,
    url: indexing.indexer.url.clone(),
    perf: perf.response,
    fee,
    seconds_behind: perf.seconds_behind,
    slashable_grt: (indexing.indexer.staked_tokens as f64 * 1e-18) as u64,
    versions_behind: indexing.versions_behind,
    zero_allocation: indexing.indexer.total_allocated_tokens == 0,
})
```

File: graph-gateway/src/network/entities.rs (hunk 2 above)
Comment (LNSD) [reply graph-gateway/src/network/entities.rs:28]: Noted down for a future PR, because I think that would require we do the same with the whole codebase, e.g., the `client_query.rs`'s candidate preparation, which is the one that has conditioned this PR

File: graph-gateway/src/network/entities.rs (hunk 4 above)
Comment (Theodus) [reply graph-gateway/src/network/entities.rs:76]: That field should be the total allocated tokens allocated for a deployment (or just the tokens associated with the allocation we select). Not the total allocated tokens across all deployments.

File: graph-gateway/src/network/entities.rs (hunk 3 above)
Comment (LNSD) [reply graph-gateway/src/network/entities.rs]: I did my best trying to untangle the original spaghetti code, but seems I missed one turn 😅 🍝 

I pushed the fix in [`348f0fd` (#706)](https://github.com/edgeandnode/gateway/pull/706/commits/348f0fdd3f219d97c3f7cc88ad80c38dc51adb9d)

File: graph-gateway/src/network/entities.rs (hunk 4 above)
Comment (LNSD) [reply graph-gateway/src/network/entities.rs]: Same here 🍝 

> 
I pushed the fix in [`348f0fd` (#706)](https://github.com/edgeandnode/gateway/pull/706/commits/348f0fdd3f219d97c3f7cc88ad80c38dc51adb9d)

File: graph-gateway/src/network/indexers_cost_model_resolver.rs (hunk 1 above)
Comment (LNSD) [reply graph-gateway/src/network/indexers_cost_model_resolver.rs]: Resolved in [`bacfaa5` (#706)](https://github.com/edgeandnode/gateway/pull/706/commits/bacfaa57051c99ee94da3e291019f91bbd10d74c)

File: graph-gateway/src/network/service.rs [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+            tokio::select! { biased;
+                update = fetch_update(&subgraph_client, &state) => {
+                    match update {
+                        Ok(network) => {
+                            eventual_writer.write(Ptr::new(network));
+                        }
+                        // If the fetch fails, log a warning and skip the update
+                        Err(err) => {
+                            tracing::warn!("network update fetch failed");
+                            tracing::debug!("network update fetch failed: {err}");
```
Comment (Theodus) [inline graph-gateway/src/network/service.rs:370]: Let's not duplicate logs like this. If we are ever looking for this log message, we probably want the full detail.
```suggestion
                            tracing::warn!("network update fetch failed: {err}");
```

Comment (Theodus) [reply graph-gateway/src/network/service.rs:370]: Also, I generally prefer the pattern of labeling the error instead of prefixing the message, like
```rust
Err(network_update_err) => tracing::warn!(%network_update_err),
```
I find it easier to filter logs this way, since it uses some structure (and it's nice to be able to scan for things like `_err`).

File: graph-gateway/tests/it_network_subgraph_client.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Theodus) [reply graph-gateway/tests/it_network_subgraph_client.rs:26]: That shouldn't be a problem anymore. The arbitrum subgraph is what we've been using in production for some time now.

**Review by Theodus — COMMENTED** [review]
Still reviewing, here's some comments so far.

File: graph-gateway/src/network/indexers_addr_blocklist.rs [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    /// If the address is in the blocklist, return [`BlocklistResult::Blocked`], otherwise return
+    /// [`BlocklistResult::Allowed`].
+    pub fn check(&self, addr: &Address) -> BlocklistResult {
+        if self.blocklist.contains(addr) {
+            BlocklistResult::Blocked
+        } else {
+            BlocklistResult::Allowed
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (LNSD) [reply graph-gateway/src/network/indexers_addr_blocklist.rs:34]: > It seems unnecessary to me to add a new file of code just to wrap a HashSet and expose contains, when a HashSet clearly represents the thing we want.

I agree that this seems unnecessary. Initially, I kept it as a simple `HashSet` instance. But I decided to encapsulate it in a `struct` for consistency with the rest of the indexer blocklist structures.

>  I think, in this case, that the cost of diving through another layer of abstraction to understand what's going on is not worth it.

I understand that this adds another layer of abstraction for the graph gateway. But we want to generalize things for the other kinds of gateways, so I believe the way to go is to keep a homogenous API (e.g., same for all the indexer blocklists) that eventually turns into an interface (i.e., trait) and allows us to mix and match the different blocklists.

----

**TL;DR:** Although this code is under the `graph-gateway` module, it is a first iteration. Once the module APIS is stable enough, it should eventually be moved to the `gateway framework` to be reused in the different gateways.

File: graph-gateway/src/network/indexers_cost_model_resolver.rs [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+use std::collections::HashMap;
+
+use thegraph_core::types::DeploymentId;
+use url::Url;
+
+use crate::{indexers, indexers::cost_models::CostModelSource};
+
+/// Error type for cost model resolution.
+#[derive(Debug, Clone, thiserror::Error)]
+enum ResolutionError {
```
[the remark this answers is not in the document]
Comment (LNSD) [reply graph-gateway/src/network/indexers_cost_model_resolver.rs]: Removed, but added a TODO to rework error reporting here once the `indexers` module bubbles up different error types and not only an `anyhow::Error`.

File: graph-gateway/tests/it_network_subgraph_client.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (LNSD) [reply graph-gateway/tests/it_network_subgraph_client.rs]: Cool! Now it works 💯 I have updated the tests to use the _Arbitrum Gateway_ URL and credentials.

File: graph-gateway/src/network/subgraph.rs [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+                }
+            ) {
+                id
+                url
+                stakedTokens
+                allocations(
+                    first: 100,
+                    orderBy: allocatedTokens, orderDirection: desc
+                    where: { status: Active }
+                ) {
```
Comment (Theodus) [inline graph-gateway/src/network/subgraph.rs:234]: This filter only makes sense if it's the first n indexer allocations per subgraph deployment. When querying for indexers, many will have more than 100 allocations (like upgrade indexer). So this would cause problems.

File: graph-gateway/src/network/entities.rs [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+    /// The subgraph's chain name.
+    ///
+    /// This information is extracted from the highest version of the subgraph deployment's
+    /// manifest.
+    pub chain: String,
+    /// The subgraph's start block number.
+    ///
+    /// This information is extracted from the highest version of the subgraph deployment's
+    /// manifest.
+    pub start_block: BlockNumber,
```
Comment (Theodus) [inline graph-gateway/src/network/entities.rs:121]: This field should not be associated with a subgraph, because it could be different at each deployment.

**Review by Theodus — APPROVED** [review]

File: graph-gateway/src/network/entities.rs (hunk 10 above)
Comment (LNSD) [reply graph-gateway/src/network/entities.rs:121]: Well... this value is used at this moment in the `client_query.rs` file, with the `manifest_min_block` name:

https://github.com/edgeandnode/gateway/blob/cef0c73d4d3be52a3220cc000ac628e6efc2cfc2/graph-gateway/src/client_query.rs#L317-L323

It is computed like this:

https://github.com/edgeandnode/gateway/blob/cef0c73d4d3be52a3220cc000ac628e6efc2cfc2/graph-gateway/src/client_query.rs#L246

This is basically what I stated in the field documentation above:

```rust
/// The subgraph's start block number.
///
/// This information is extracted from the highest version of the subgraph deployment's
/// manifest.
pub start_block: BlockNumber,
```

The only change here is that we are pre-computing it.

Comment (Theodus) [reply graph-gateway/src/network/entities.rs:121]: ah, ok. I think I got confused between this field and the other one in the subgraph struct when scanning through.

**Comment by LNSD** [comment]
Part of https://github.com/edgeandnode/gateway/issues/791
